client/subscriptions.go

```diff
@@ -0,0 +1,159 @@
+package client
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/EventStore/EventStore-Client-Go/internal/protoutils"
+	"github.com/EventStore/EventStore-Client-Go/position"
+	system_metadata "github.com/EventStore/EventStore-Client-Go/systemmetadata"
+
+	"github.com/EventStore/EventStore-Client-Go/messages"
+	api "github.com/EventStore/EventStore-Client-Go/protos/streams"
+)
+
+type SubscriptionEvent struct {
+	EventAppeared     *messages.RecordedEvent
+	Dropped           *SubscriptionDropped
+	CheckPointReached *position.Position
+}
+
+type SubscriptionDropped struct {
+	Error error
+}
+
+type request struct {
+	close   bool
+	channel chan *SubscriptionEvent
+}
+
+type Subscription struct {
+	client  *Client
+	id      string
+	inner   api.Streams_ReadClient
+	channel chan request
+}
+
+func NewSubscription(client *Client, inner api.Streams_ReadClient, id string) *Subscription {
+	channel := make(chan request)
+
+	// It is not safe to consume a stream in different goroutines. This is why we only consume
+	// the stream in a dedicated goroutine.
+	//
+	// Current implementation doesn't terminate the goroutine. When a subscription is dropped,
+	// we keep user requests coming but will always send back a subscription dropped event.
+	// This implementation is simple to maintain while letting the user sharing their subscription
+	// among as many goroutines as they want.
+	go func() {
+		closed := false
+
+		for {
+			req := <-channel
+
+			if req.close {
+				closed = true
+				inner.CloseSend()
+				continue
+			}
+
+			if closed {
+				req.channel <- &SubscriptionEvent{
+					Dropped: &SubscriptionDropped{
+						Error: fmt.Errorf("subscription has been dropped"),
+					},
+				}
+
+				continue
+			}
+
+			result, err := inner.Recv()
+			if err != nil {
+				log.Printf("[error] subscription has dropped. Reason: %v", err)
+
+				dropped := SubscriptionDropped{
+					Error: err,
+				}
+
+				req.channel <- &SubscriptionEvent{
+					Dropped: &dropped,
+				}
+
+				closed = true
+
+				continue
+			}
+
+			switch result.Content.(type) {
+			case *api.ReadResp_Checkpoint_:
+				{
+					checkpoint := result.GetCheckpoint()
+					position := position.Position{
+						Commit:  checkpoint.CommitPosition,
+						Prepare: checkpoint.PreparePosition,
+					}
+
+					req.channel <- &SubscriptionEvent{
+						CheckPointReached: &position,
```

Comment:
Now that checkpoints are handled differently, should we remove the `CheckpointReached` func on the `NewSubscriptionFilterOptions` as that callback does nothing now.

Reply:
done